Add ST_GraphAnalysis

agouge opened the pull request:
This is a really huge PR. Next time, I'll break it into smaller chunks.

Some things to note:
1. Remove `ST_ClosenessCentrality`. This has been replaced by `ST_GraphAnalysis` which is faster anyway.
2. Add [`GraphCreator`s](https://github.com/agouge/gdms-topology/tree/GH/src/main/java/org/gdms/gdmstopology/graphcreator) (and remove `GraphLoaderUtilities`). This simplifies loading a graph from a `DataSource`.
3. Add [function parsers](https://github.com/agouge/gdms-topology/tree/GH/src/main/java/org/gdms/gdmstopology/parse) which simplify multiple function signatures for SQL functions.
4. Add [AbstractExecutorFunctionHelper](https://github.com/agouge/gdms-topology/blob/GH/src/main/java/org/gdms/gdmstopology/centrality/AbstractExecutorFunctionHelper.java). This helps automate a lot of  the tasks that are common to the implementations of SQL executor functions.
5. Add `ST_GraphAnalysis`. For now, we compute betweenness and closeness. Since the progress monitor is not currently working in OrbisGIS, I made my own progress monitor which prints output to the console.
6. Change the bundle category to "Network Analysis". This should facilitate locating the plugin within the plugin manager.

This version of GDMS-Topology uses [GraphHopper-SNA](https://github.com/agouge/graphhopper-sna) which currently uses [GraphHopper](https://github.com/graphhopper/graphhopper).

agouge committed:
Added graphhopper, graphhopper-sna, trove dependencies.

agouge committed:
Calculate closeness using graphhopper(-sna).

agouge committed:
Add methods for loading GH graphs from data sets.

agouge committed:
Update pom.

agouge committed:
Make ST_ClosenessCent... parse optional arguments.

agouge committed:
Fix pom.

agouge committed:
Fix minor errors.

agouge committed:
Update unit tests.

agouge committed:
Rename Undirected... to Unweighted...

agouge committed:
Fix method name.

agouge committed:
Remove jgrapht-sna centrality computations.

agouge committed:
Update dependencies in pom.

agouge committed:
Update gitignore.

agouge committed:
Fix some dependencies.

agouge committed:
Compute closeness on weighted graphs.

agouge committed:
Add tests for Nantes in comments.

agouge committed:
Add classes for parsing SQL function arguments.

agouge committed:
Revise how SQL function parameters are parsed.

agouge committed:
Add default values and update function signatures.

agouge committed:
Clean up code.

agouge committed:
Change graph loader parameter order.

agouge committed:
Add class for calculating shortest paths using GH.

agouge committed:
Make ST_ShortestPath support unweighted graphs.

agouge committed:
Modify the tests for ST_ShortestPath.

agouge committed:
Update debugging output of ST_ShortestPath.

agouge committed:
Add method for printing a DiskBufferDriver.

agouge committed:
Make GRAPH_TYPE_ERROR public.

agouge committed:
Fix typo: Remove extra comma.

agouge committed:
Add GH to pom and change scope of GH-SNA.

agouge committed:
Merge branch 'GH' into shortestPath

agouge committed:
Refactor SP String messages into local fields.

agouge committed:
Refactor CC String messages into local fields.

agouge committed:
Merge branch 'refactor' into GH

Conflicts:
	src/main/java/org/gdms/gdmstopology/function/ST_ConnectedComponents.java

agouge committed:
Merge branch 'GH' into shortestPath

agouge committed:
Refactor SP String messages into local fields.

agouge committed:
Adjust description of ST_ShortestPath.

agouge committed:
Merge branch 'refactor' into shortestPath

Conflicts:
	src/main/java/org/gdms/gdmstopology/function/ST_ShortestPath.java

agouge committed:
Update javadoc.

agouge committed:
Refactor closeness String msgs into local fields.

agouge committed:
Refactor ShortestPath function description.

agouge committed:
Merge branch 'refactor' into shortestPath

Conflicts:
	src/main/java/org/gdms/gdmstopology/function/ST_ClosenessCentrality.java
	src/main/java/org/gdms/gdmstopology/function/ST_ShortestPath.java

agouge committed:
Take GH changes into account.

agouge committed:
Rename some AllWeightsOne -> Unweighted.

agouge committed:
Fix some formatting and imports.

agouge committed:
Rename test file.

agouge committed:
Update .gitignore.

agouge committed:
Refactor into centrality package.

agouge committed:
Update license.

We will need to update all the license headers later.

agouge committed:
Move parsers to their own package.

agouge committed:
Make parser give more informative output.

agouge committed:
Update pom

agouge committed:
Update formatting

agouge committed:
No need to extend GraphAnalysis class.

agouge committed:
Add GraphAnalyzers.

agouge committed:
Fix version numbers in pom.

agouge committed:
Update name, url, and bundle category in pom.

agouge committed:
Update local variable name.

agouge committed:
Refactor analyzers to load graphs themselves.

agouge committed:
Fix javadoc formatting.

agouge committed:
Move graph prep from analyzer to superclass.

agouge committed:
Move graph prep to GraphCreator classes.

agouge committed:
Move graph creators to their own package.

agouge committed:
Delete GraphLoaderUtilities.

This has been replaced by GraphCreator.

agouge committed:
Rename and move GraphCentralityUtilities ->
ClosenessCentralityUtilities.

agouge committed:
Refactor abstract executor functionality to helper

Most of the work done in GraphAnalyzer could be pushed up to a
superclass called AbstractExecutorFunctionHelper which will be reused by
other SQL functions extending AbstractExecutorFunction.

agouge committed:
Make writeToTable give the correct output suffix.

agouge committed:
Major refactor of ST_ClosenessCentrality.

agouge committed:
Update javadoc to be more abstract.

agouge committed:
Simplify computeAll() for ClosenessComputers.

agouge committed:
Add prepareAnalyzer() and move computeAll() up.

agouge committed:
Rename graphType -> orientation.

agouge committed:
Update javadoc.

agouge committed:
Refactor ST_GraphAnalysis function signatures.

agouge committed:
Fix minor documentation.

agouge committed:
Update javadoc/usage guidelines: ST_GraphAnalysis.

agouge committed:
Oops, forgot ArrayConcatenator.

This file should have been included in commit
011d79e3f4e9529059000cac5b83fe511f54f04a.

agouge committed:
Verify the start, end and weight indices.

agouge committed:
Add GraphCreator tests.

agouge committed:
Change the backupDir to fix failing tests Jenkins.

agouge committed:
Merge branch 'master' into GH

Conflicts:
	src/test/java/org/gdms/gdmstopology/TopologySetupTest.java
	src/test/java/org/gdms/gdmstopology/function/CentralityTest.java

Deleted CentralityTest.java.

agouge committed:
Wrap a GraphHopper-SNA pm to an OrbisGIS pm.

agouge committed:
Add level graph preparation for closeness.

agouge committed:
Remove ST_ClosenessCentrality.

It is slow anyway, and is included in ST_GraphAnalysis.

agouge committed:
Update due to upstream changes.

agouge committed:
No longer need the level graph (was for closeness)

agouge committed:
Change bundle category to "Network Analysis".

nicolas-f commented on src/main/java/org/gdms/gdmstopology/graphcreator/GraphCreator.java:
```diff
@@ -0,0 +1,249 @@
+import com.graphhopper.storage.GraphStorage;
+import com.graphhopper.storage.RAMDirectory;
+import java.util.logging.Level;
+import java.util.logging.Logger;
```
We are using LOG4J for loggin, java.util.logging will not be visible in OrbisGIS log panel.

agouge replied:
OK, I'll make the conversion.

agouge replied:
Done

agouge committed:
Fix where SP metadata comes from.

agouge committed:
Merge branch 'master' into GH

I delete DemoCentralityIndices as in c062d0382db. I also keep
ST_GraphAnalysis in the Activator and lower the version of OSGi from
5.0.0 to 4.3.1.

In the next commit, I'll restore the version of ST_ShortestPath written
by Erwan that he just fixed in c062d0382db.

Conflicts:
	pom.xml
	src/main/java/org/gdms/gdmstopology/Activator.java
	src/main/java/org/gdms/gdmstopology/demo/DemoCentralityIndices.java
	src/main/java/org/gdms/gdmstopology/function/ST_ShortestPath.java

agouge committed:
Restor Erwan's version of ST_ShortestPath.

I had written a different version of this function that didn't quite
work correctly. I would like to refactor this function later to make it
easier to read and debug.

agouge committed:
Move from Java logging to log4j.
